src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxInputNode.cs

```diff
@@ -61,35 +67,42 @@ public void SaveStateAndFree(ImmutableSegmentedDictionary<object, IStateTable>.B
                 tables[_owner] = _transformTable.ToImmutableAndFree();
             }
 
-            public void VisitTree(Lazy<SyntaxNode> root, EntryState state, SemanticModel? model, CancellationToken cancellationToken)
+            public void VisitTree(Lazy<SyntaxNode> root, EntryState state, int syntaxTreeIndex, IncrementalGeneratorRunStep? inputStep, GeneratorRunStateTable.Builder runStateTableBuilder, SemanticModel? model, CancellationToken cancellationToken)
             {
                 if (state == EntryState.Removed)
                 {
                     // mark both syntax *and* transform nodes removed
-                    _filterTable.RemoveEntries();
-                    _transformTable.RemoveEntries();
+                    _filterTable.TryRemoveEntries(TimeSpan.Zero, ImmutableArray<(IncrementalGeneratorRunStep, int)>.Empty, out ImmutableArray<SyntaxNode> removedNodes);
+
+                    for (int i = 0; i < removedNodes.Length; i++)
```

Comment:
>for (int i = 0; i < removedNodes.Length; i++) [](http://example.com/codeflow?start=20&length=45)

I'm confused by this loop. Are we creating more removed entries than before this change?

Reply:
Yes, beforehand we did not accurately create enough "Removed" entries in the table, but since entries with the `Removed` state were ignored in later stages anyway (only kept in very specific scenarios), the bug was unnoticed.